pypsa/networkclustering.py

```diff
@@ -898,6 +898,97 @@ def hac_clustering(
     )
 
 
+################
+# Cluserting based on Modularity (on electrical parameters of the network)
+def busmap_by_greedy_modularity(network, n_clusters, buses_i=None):
+    """
+    Create a busmap according to Clauset-Newman-Moore greedy modularity
+    maximization [1].
+
+    Parameters
+    ----------
+    network : pypsa.Network
+    n_clusters : int
+        Final number of clusters desired.
+    buses_i: None | pandas.Index, default=None
+        Subset of buses to cluster. If None, all buses are considered.
+
+    Returns
+    -------
+    busmap : pandas.Series
+        Mapping of network.buses to clusters (indexed by
+        non-negative integers).
+
+    References
+    ----------
+    .. [1] Clauset, A., Newman, M. E., & Moore, C.
+       "Finding community structure in very large networks."
+       Physical Review E 70(6), 2004.
+    """
+
+    if buses_i is None:
+        buses_i = network.buses.index
+
+    network.calculate_dependent_values()
+
+    lines = network.lines.query("bus0 in @buses_i and bus1 in @buses_i")
+    lines = (
+        lines.loc[:, ["bus0", "bus1"]]
+        .assign(weight=network.lines.s_nom / abs(lines.r + 1j * lines.x))
+        .set_index(["bus0", "bus1"])
+    )
+
+    G = nx.Graph()
+    G.add_nodes_from(buses_i)
+    G.add_edges_from((u, v, dict(weight=w)) for (u, v), w in lines.itertuples())
+
+    communities = nx.community.greedy_modularity_communities(
+        G, best_n=n_clusters, cutoff=n_clusters, weight="weight"
+    )
+    busmap = pd.Series(buses_i, buses_i)
```

Comment:
Is communities a dict?

Reply:
should be a list: https://github.com/networkx/networkx/blob/main/networkx/algorithms/community/modularity_max.py#L294